[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39410 [vulkan] Remove unused variables
* #39384 [vulkan] Shaders caching
* **#39282 [vulkan] jit passes for vulkan conv2 prepack and fuse with clamp**

Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

IvanKobzarev committed:
[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp

[ghstack-poisoned]

IvanKobzarev committed:
rebase on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"




[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanConvolution.cpp:
```diff
@@ -66,14 +66,14 @@ ContextConv2D create(
   const auto dilation_expanded =
       expand_param_if_needed(dilation, "dilation", 2);
-  const Tensor weight_nchw = weight.contiguous();
+  Tensor weight_nchw = weight.contiguous();
```
Does a call to contiguous() return NCHW format?  This always confuses me and for some reason I always default to the assumption that contiguous() returns a tensor laid out in memory contiguously while the elements can be in any order as long as they are all adjacent without any holes.

Anyhow, do you depend on the memory to be NCHW or do you just want the values to be adjacent in memory?  We have noticed in the past that it is very easy to cause unnecessary memory copies with contiguous() calls so if performance matters, you want to vett all usage of this function to make sure it is only doing the minimal work necessary.

kimishpatel replied:
Ideally we can call this out explicitly in each contiguous call. Like `contiguous(MemoryFormat::Contiguous)`. Less confusion.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanConvolution.cpp:
```diff
@@ -66,14 +66,14 @@ ContextConv2D create(
   return ContextConv2D{
-      at::native::vulkan_convolution_prepack_weights(weight),
+      groups == 1 ? at::native::vulkan_convolution_prepack_weights(weight_nchw)
+                  : std::move(weight_nchw.vulkan()),
```
Do we have an assert somehwere that weight_nchw.is_vulkan() is true?

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanConvolution.cpp:
```diff
@@ -66,14 +66,14 @@ ContextConv2D create(
-       weight_nchw.sizes()[1],
-       weight_nchw.sizes()[2],
-       weight_nchw.sizes()[3]},
+      weight_sizes,
```
You could still pass this inline with the aggregate initializer list as you were doing, after having factored out weight_nchw.sizes().  Not that it should matter if the compiler is optimizing out the copy, as it should, but generally speaking using temporaries is a safer bet when possible.

This is merely a suggestion.  What you have is perfectly fine.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -255,7 +250,7 @@ void conv2d_depthwise(
   output.image()->bindStorageImage(descriptorSet, 0);
   input.image()->bindShaderRead(descriptorSet, 1);
-  kernel.image()->bindShaderRead(descriptorSet, 2);
+  weight.image()->bindShaderRead(descriptorSet, 2);
```
Do we have an assert or a runtime check that these are is_image()?

AshkanAliabadi reviewed (commented):
Got half-way through.  Will review the rest later.

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -279,6 +274,44 @@ void conv2d_depthwise(
+    const VulkanTensor& input,
+    const VulkanTensor& weight,
+    const c10::optional<float*> bias,
+    const Conv2DParams params,
```
Consider passing  by const-ref.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanConvolution.cpp:
```diff
@@ -66,14 +66,13 @@ ContextConv2D create(
   return ContextConv2D{
-      at::native::vulkan_convolution_prepack_weights(weight),
+      groups == 1 ? at::native::vulkan_convolution_prepack_weights(weight_nchw)
+                  : std::move(weight_nchw.vulkan()),
```
I'm not sure a 'move' is necessary since weight_nchw.vulkan() is already an r-value.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -279,6 +274,44 @@ void conv2d_depthwise(
+    VulkanTensor& output,
+    const VulkanTensor& input,
+    const VulkanTensor& weight,
+    const c10::optional<float*> bias,
```
You could consider c10::optional<const float*> since the memory location bias is pointing to is not supposed to change.  If you run into compilation issues because c10::optional does not provide implicit conversion constructors to convert from float* to const float* then please skip this suggestion.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -279,6 +274,44 @@ void conv2d_depthwise(
+      bufferFromOptionalHostData(bias, conv2d_biasBufferSize(params.OC)),
+      params,
+      output_min,
+      output_max);
```
I recognize this might be difficult to do in eager mode but for performance reasons we want to upload weights and biases to GPU memory ahead of time, and preferably do so with a significant gap, ideally as a one time operation at initialization time.  This usage here is going to act as a sync point otherwise hurting performance.

If weight prepacking is only covering memory reordering and conversion, and weight uploads to GPU memory is still done inline like here, do you envision us having a path to do the aforementioned one-time weight & bias upload?

IvanKobzarev replied:
Currently prepacking is folded with https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/prepack_folding.cpp to the start of `forward()` method, (That's my current understanding, I only read it briefly). I think we can play around that folding to move our prepacking to other places of the graph, probably it's loading if it is possible

kimishpatel replied:
So prepack folding removes all prepacking ops and inserts the outputs of prepacking op as an attribute of the top level module. The attributes are loaded only once so those are done only once.

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

AshkanAliabadi commented on torch/csrc/jit/passes/vulkan_rewrite.cpp:
```diff
@@ -0,0 +1,240 @@
+#include <ATen/core/jit_type.h>
```
@kimishpatel Can you please review this file?

kimishpatel commented on aten/src/ATen/test/vulkan_test.cpp:
```diff
@@ -507,7 +507,7 @@ TEST(VulkanTest, conv2dPrepack) {
       output_max);
   auto tv_out_prepack_ivalues =
-      callOpByName("vulkan::conv2d_clamp_run", "", tv_in, prepack[0]);
+      callOpByName("vulkan_prepack::conv2d_clamp_run", "", tv_in, prepack[0]);
```
I think decision regarding these are already made but I am wondering why are the ops no in plain `prepacked` namespace instead of `vulkan_prepack`. I am asking this because we explicitly removed `xnnpack` from the namespace for such prepacking based ops, precisely so that we have the same op that dispatch to different implementation depending on tensor type.

IvanKobzarev replied:
Ops have the same name as xnnpack prepacking and for prepack it is called on CPU tensor, thats why I decided to use namespace split with the same naming. Probably conv2d_clamp_run can be in  `prepacked`  namespace as xnnpack,dispatching by DeviceType, but for prepack operation it is on the same DeviceType.

kimishpatel commented on torch/csrc/jit/passes/vulkan_rewrite.cpp:
```diff
@@ -0,0 +1,240 @@
+  rewriter.runOnGraph(graph);
+}
+
+bool isClampFusable(
```
This function I think is the duplicate of the original one, right? If so can we move this to some common file, such as https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/graph_rewrite_helper.cpp or something else.

IvanKobzarev replied:
Yes, as xnnpack has `#ifdef XNNPACK` - I completely copy pasted it not to have any compilation dependencies.

kimishpatel commented on torch/csrc/jit/passes/vulkan_rewrite.h:
```diff
@@ -0,0 +1,15 @@
+TORCH_API c10::optional<script::Module> vulkanOptimizeForMobile(
+    const script::Module& module);
+} // namespace jit
+} // namespace torch
```
Also can you add some tests for graph rewriting? You can follow this one: https://github.com/pytorch/pytorch/blob/master/test/test_xnnpack_integration.py#L380. This is just to make sure we are indeed doing the right transformation. Although the function correctness check might not be possible in CI we can at least check graph rewrite.

kimishpatel reviewed (changes requested):
Overall looks fine to me. Ideally we combine optimizeForMobile and VulkanOptimize for them to share common optimization passes and have a flag passes that does vulkan specific rewrites if that is what we want, but maybe we can do this in later PR.

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

kimishpatel commented on test/test_vulkan.py:
```diff
@@ -0,0 +1,131 @@
+        pattern_count_map = {"Tensor = aten::conv2d": -1,
+                             "vulkan_prepack::conv2d_clamp_prepack": 1,
+                             "vulkan_prepack::conv2d_clamp_run": 1}
+        TestVulkanRewritePass.validate_transformed_module(Conv2D(), pattern_count_map, data_shape)
```
Check prepack removal pass as well?

kimishpatel commented on test/test_vulkan.py:
```diff
@@ -0,0 +1,131 @@
+                             "Tensor = aten::conv2d": -1,
+                             "vulkan_prepack::conv2d_clamp_prepack": 1,
+                             "vulkan_prepack::conv2d_clamp_run": 1}
+        TestVulkanRewritePass.validate_transformed_module(Conv2DRelu(), pattern_count_map, data_shape)
```
Check clamp fusion pass? Same for hardtanh fusion.

kimishpatel reviewed (approved):
I am accepting but for graph rewriting test you have, can you update the tests so that prepack removal and relu/hardtanh fusion are tested as well. Should be simple as it requires just passing additional parameters to validate.. function.

kimishpatel commented:
Also it seems that graph rewrite tests are failing.

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] jit passes for vulkan conv2 prepack and fuse with clamp"


Differential Revision: [D21962424](https://our.internmc.facebook.com/intern/diff/D21962424)

[ghstack-poisoned]
